### PR TITLE
Regularize mask handling for attn_mask and key_padding_mask

### LINKED ISSUE #92554: Major bug in Transformers' masks
### 🐛 Describe the bug

In torch.nn.functional, line 5174 in current code:
image

`attn_mask = attn_mask.masked_fill(key_padding_mask, float("-inf"))`

Fills `-inf` value wherever `key_padding_mask` is True. However, `key_padding_masks` actually contains -inf values instead of True values. So the padding mask is actually ignored. Replacing by:

`attn_mask = attn_mask.masked_fill(key_padding_mask < 0, float("-inf")`

Gives the correct expected mask.

### Versions

Current github code.  (master branch, Jan 18, 2023)

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @bhosmer @cpuhrsch @erichan1 @drisspg @saketh-are

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary:
Regularize mask handling for attn_mask and key_padding_mask
* Update documentation to remove reference to byte masks (which were deprecated long ago)
* Introduce check and warn about deprecation if attn_mask and key_padding_mask types mismatch
* Convert all masks to float before combining
* Combine by adding

Test Plan: sandcastle & github CI

Differential Revision: D42653215

### CODE REVIEW

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -1037,7 +1037,7 @@ def forward(
             See "Attention Is All You Need" for more details.
         key_padding_mask: If specified, a mask of shape :math:`(N, S)` indicating which elements within ``key``
             to ignore for the purpose of attention (i.e. treat as "padding"). For unbatched `query`, shape should be :math:`(S)`.
-            Binary and byte masks are supported.
+            Binary and floati masks are supported.
```
Comment (ngimel) [inline torch/nn/modules/activation.py:1040]:
```suggestion
            Binary and float masks are supported.
```

File: torch/nn/functional.py [hunk 2]
```diff
@@ -5058,6 +5054,18 @@ def multi_head_attention_forward(
         if _kpm_dtype != torch.bool and not torch.is_floating_point(key_padding_mask):
             raise AssertionError(
                 "only bool and floating types of key_padding_mask are supported")
+        if attn_mask is not None:
+            if _kpm_dtype != attn_mask.dtype:
+                warnings.warn("Support for mismatched attn_mask and kay_padding_mask in nn.MultiheadAttention is deprecated. Use same type for both instead.")
```
Comment (drisspg) [inline torch/nn/functional.py:5059]: nit: key

**Comment by ngimel** [comment]
I'm a bit worried that we need to put in key_padding_mask conversion in so many places (and btw I think it's also needed in nn/modules/transformer.py in TransformerEncoder module), could we create a helper function that would be called from everywhere if we don't want a more serious refactor? 
There is a small perf impact from conversion but I think that's fine. 
What about tests?

**Comment by mikekgfb** [comment]
> I'm a bit worried that we need to put in key_padding_mask conversion in so many places (and btw I think it's also needed in nn/modules/transformer.py in TransformerEncoder module), could we create a helper function that would be called from everywhere if we don't want a more serious refactor?

We had a more serious refactor, but it collided with landing sdpa() which was a priority.  Totally agree it is desirable.  Will be good to find somebody to do that.

> could we create a helper function that would be called from everywhere if we don't want a more serious refactor?

Can we do this as a follow-up task?  It's getting unwieldy and I'd like to land a solution.  (I'd desist from fixing transformer encoder and land that concurrently with the helper function update.)

> What about tests?

Transformers and Attention are already tested, right?  So, simplifying the implementation shouldn't require new tests?
(I'll create a task to track creating more tests, because clearly we missed some combinations.)

**Comment by ngimel** [comment]
But existing tests didn't uncover this bug, so we need new ones?

**Comment by mikekgfb** [comment]
Refactored the testing and implemented canonical mask conversion across F.MHA, nn.MHA, and nn.TransformerEncoder.  Opens the path for a bigger refactor to validate and combine masks as a common building block in a future update.

> But existing tests didn't uncover this bug, so we need new ones?

I added a suitable test.

File: torch/nn/functional.py [hunk 3]
```diff
@@ -4910,6 +4910,41 @@ def _mha_shape_check(query: Tensor, key: Tensor, value: Tensor,
 
     return is_batched
 
+def canonical_mask(
```
Comment (cpuhrsch) [inline torch/nn/functional.py:4913]: Can we add an underscore since these are not public functions?

That is `_canonical_mask` instead of `canonical_mask`

File: torch/nn/functional.py [hunk 4]
```diff
@@ -4910,6 +4910,41 @@ def _mha_shape_check(query: Tensor, key: Tensor, value: Tensor,
...
+                    "is deprecated. Use same type for both instead."
+                )
+        if not _mask_is_float:
+            mask = (
+                torch.zeros_like(mask, dtype=target_type)
+                .masked_fill_(mask, float("-inf"))
+            )
+    return mask
+
+def none_or_dtype(input: Optional[Tensor]) -> Optional[DType]:
```
Comment (cpuhrsch) [inline torch/nn/functional.py:4941]: Same here

**Review by cpuhrsch — CHANGES_REQUESTED** [review]

File: torch/nn/functional.py (hunk 3 above)
Comment (mikekgfb) [reply torch/nn/functional.py:4913]: Done

File: torch/nn/functional.py (hunk 4 above)
Comment (mikekgfb) [reply torch/nn/functional.py:4941]: Done

Comment (mikekgfb) [reply torch/nn/functional.py:4941]: Approval please?

File: test/test_transformers.py [hunk 5]
```diff
@@ -1454,6 +1454,25 @@ def test_is_causal_gpu(self):
...
+    def test_mask(self):
+        def ones_tensor(*shape):
+            return torch.ones(shape, dtype=torch.float32)
+        S, L, E, H = 1, 2, 4, 1
+        qkv = ones_tensor(S, L, E)
+
+        mha = nn.MultiheadAttention(E, H)
+        mha.in_proj_weight = Parameter(torch.ones((E * 3, E)))
+        mha.out_proj.weight = Parameter(torch.ones((E, E)))
+        expected = torch.ones(size=(S, L, E)) * 16
```
Comment (drisspg) [inline test/test_transformers.py:1466]: nit: is this used?
Are is this test just to ensure that kpm and am can't both be set? if so I think the test name should somewhat reflect that

**Review by ngimel — APPROVED** [review]

File: torch/nn/modules/transformer.py [hunk 6]
```diff
@@ -720,6 +723,7 @@ def _ff_block(self, x: Tensor) -> Tensor:
 
 
 def _get_clones(module, N):
+    # FIXME: copy.deepcopy() is not defined on nn.module
```
Comment (drisspg) [inline torch/nn/modules/transformer.py:726]: Is there an issue for this?

**Review by drisspg — APPROVED** [review]
